src/storage.ts

```diff
@@ -1,13 +1,59 @@
-type IStorage = {
-    count: number;
-};
-
-const defaultStorage: IStorage = {
-    count: 0,
-};
-
-export const storage = {
-    get: (): Promise<IStorage> =>
-        chrome.storage.sync.get(defaultStorage) as Promise<IStorage>,
-    set: (value: IStorage): Promise<void> => chrome.storage.sync.set(value),
-};
+import { writable, type Writable } from 'svelte/store';
+
+/**
+ * Creates a persistent Svelte store backed by Chrome's sync storage.
+ * @template T The type of the store's value
+ * @param key The key to use in Chrome's storage
+ * @param initialValue The initial value of the store
+ * @returns A writable Svelte store
+ */
+export function persistentStore<T>(key: string, initialValue: T): Writable<T> {
+    const store = writable(initialValue);
+    let storeValues: T[] = [];
+    let chromeValues: T[] = [];
+
+    function watchStore() {
+        store.subscribe((value) => {
+            // Prevent circular updates
+            if (chromeValues && value === chromeValues[0]) {
+                chromeValues.shift();
+                return;
+            }
+
+            storeValues.push(value);
+            chrome.storage.sync.set({ [key]: value });
+        });
+    }
+
+    function watchChrome() {
+        chrome.storage.sync.onChanged.addListener((changes) => {
+            if (!(Object.hasOwn(changes, key))) return;
+
+            const value = changes[key].newValue as T;
+            console.log(`watchChrome got ${value}`);
+            if (storeValues && value === storeValues[0]) {
+                storeValues.shift();
+                return;
+            }
+
+            chromeValues.push(value);
+            store.set(value);
+        });
+    }
+
+    // Initialize the store with the value from Chrome storage
+    chrome.storage.sync.get(key).then((result) => {
+        let value = Object.hasOwn(result, key) ? result[key] : initialValue;
+        if (!Object.hasOwn(result, key)) {
+            console.log(`Persistent store: couldn't find key [${key}] in chrome storage. Default to initial value [${initialValue}]`)
+        }
+        chromeValues.push(value);
```

Comment:
```suggestion
```
It looks like we don't need to add the value to the array for synchronization because `watchStore` is called later and we will skip removing this value from `chromeValues`.

Reply:
On the first call to a `store.subscribe(foo)` function, `foo` is called. It's always called with currently-existing value.
This will cause a double-update. (chrome -> svelte -> chrome)